File: libjs/src/keys/ECDSAKey.ts
import {Buffer} from "node:buffer";

// ECDSA keys implementation using nodeJS
import type { IHiveKey } from "./IHiveKey.ts";
import crypto from "node:crypto";

export class ECDSAKey implements IHiveKey {
    privKey: crypto.KeyObject | undefined
    pubKey: crypto.KeyObject | undefined

    constructor() {
    }


    // exportPrivate returns the encoded private key if available
    public exportPrivate(): string {
        if (!this.privKey) {
            throw ("private key not created or imported")
        }
        const privPEM = this.privKey.export({
            format: "pem", // pem, der or jwk
            type: "pkcs8",  // or sec1
        })
        return privPEM.toString()
    }

    // exportPublic returns the encoded public key if available
    public exportPublic(): string {
        if (!this.pubKey) {
            throw ("public key not created or imported")
        }
        const pubPEM = this.pubKey.export({
            format: "pem", // pem, der or jwk
            type: "spki",
        })
        return pubPEM.toString()
    }

    // importPrivate reads the key-pair from the encoded private key
    // This throws an error if the encoding is not a valid key
    public importPrivate(privateEnc: string): IHiveKey {
        // cool! crypto does all the work
        this.privKey = crypto.createPrivateKey(privateEnc)
        this.pubKey = crypto.createPublicKey(privateEnc)
        return this
    }

    // importPublic reads the public key from the encoded data.
    // This throws an error if the encoding is not a valid public key
    public importPublic(publicEnc: string): IHiveKey {
        this.pubKey = crypto.createPublicKey(publicEnc)
        return this
    }


    // initialize generates a new key set using its curve algorithm
    public initialize(): IHiveKey {
        const kp = crypto.generateKeyPairSync("ec", {
            namedCurve: "secp256k1"
        })
        this.privKey = kp.privateKey
        this.pubKey = kp.publicKey
        return this
    }

    // return the signature of a message signed using this key
    // this requires a private key to be created or imported
    public sign(message: Buffer): Buffer {
        if (!this.privKey) {
            throw ("private key not created or imported")
        }
        const sigBuf = crypto.sign("sha256", message, this.privKey)
        return sigBuf
    }

    // verify the signature of a message using this key's public key
    // this requires a public key to be created or imported
    // returns true if the signature is valid for the message
    public verify(signature: Buffer, message: Buffer): boolean {
        if (!this.pubKey) {
            throw ("public key not created or imported")
        }
        const isValid = crypto.verify("sha256", message, this.pubKey, signature)
        return isValid
    }
}

File: lib/keys/EcdsaKey.go
package keys

import (
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/sha256"
	"crypto/x509"
	"fmt"
	"os"
	"reflect"
)

// EcdsaKey implements the secp256k ECDSA key.
// This implements the IHiveKeys interface.
type EcdsaKey struct {
	// Reuse RSA functionality
	RsaKey
}

// ImportPrivate reads the key-pair from the PEM private key
// and determines its key type.
// This returns an error if the PEM is not a valid key.
func (k *EcdsaKey) ImportPrivate(privatePEM string) (err error) {
	derBytes, err := k._importDer(privatePEM)
	_ = err
	rawPrivateKey, err := x509.ParsePKCS8PrivateKey(derBytes)
	if err != nil {
		return err
	}

	privKey, valid := rawPrivateKey.(*ecdsa.PrivateKey)
	if !valid {
		keyType := reflect.TypeOf(k.pubKeyPtr)
		return fmt.Errorf("not an ECDSA private key. It looks to be a '%s'", keyType)
	}
	k.privKeyPtr = privKey
	k.pubKeyPtr = &privKey.PublicKey

	return err
}

// ImportPrivateFromFile loads public/private key pair from PEM file
// and determines its key type.
func (k *EcdsaKey) ImportPrivateFromFile(pemPath string) (err error) {
	pemEncodedPriv, err := os.ReadFile(pemPath)
	if err != nil {
		return err
	}
	err = k.ImportPrivate(string(pemEncodedPriv))
	return err
}

// ImportPublic reads the public key from the PEM data.
// This returns an error if the PEM is not a valid public key
//
// publicPEM must contain either a PEM encoded string, or its base64 encoded content
func (k *EcdsaKey) ImportPublic(publicPEM string) (err error) {
	derBytes, err := k._importDer(publicPEM)
	_ = err
	k.pubKeyPtr, err = x509.ParsePKIXPublicKey(derBytes)
	k.privKeyPtr = nil
	_, valid := k.pubKeyPtr.(*ecdsa.PublicKey)
	if !valid {
		keyType := reflect.TypeOf(k.pubKeyPtr)
		return fmt.Errorf("not an ECDSA public key. It looks to be a '%s'", keyType)
	}
	return err
}

// ImportPublicFromFile loads ECDSA public key from PEM file
func (k *EcdsaKey) ImportPublicFromFile(pemPath string) (err error) {
	pemEncodedPub, err := os.ReadFile(pemPath)
	if err != nil {
		return err
	}
	err = k.ImportPublic(string(pemEncodedPub))
	return err
}

// Initialize generates a new key
func (k *EcdsaKey) Initialize() IHiveKey {
	curve := elliptic.P256()
	privKey, err := ecdsa.GenerateKey(curve, rand.Reader)
	if err != nil {
		panic("unable to create ECDSA key")
	}
	k.privKeyPtr = privKey
	k.pubKeyPtr = &privKey.PublicKey
	return k
}

// KeyType returns this key's type, eg ecdsa
func (k *EcdsaKey) KeyType() KeyType {
	return KeyTypeECDSA
}

// Sign returns the signature of a message signed using this key
// This signs the SHA256 hash of the message
// this requires a private key to be created or imported
func (k *EcdsaKey) Sign(msg []byte) (signature []byte, err error) {
	msgHash := sha256.Sum256(msg)
	privKey := k.privKeyPtr.(*ecdsa.PrivateKey)

	signature, err = ecdsa.SignASN1(rand.Reader, privKey, msgHash[:])
	return signature, err
}

// Verify the signature of a message using this key's public key
// This verifies using the SHA256 hash of the message.
// this requires a public key to be created or imported
// returns true if the signature is valid for the message
func (k *EcdsaKey) Verify(msg []byte, signature []byte) (valid bool) {
	msgHash := sha256.Sum256(msg)
	pubKey := k.pubKeyPtr.(*ecdsa.PublicKey)
	valid = ecdsa.VerifyASN1(pubKey, msgHash[:], signature)
	return valid
}

// NewEcdsaKey creates and initialize a ECDSA key
func NewEcdsaKey() IHiveKey {
	nk := &EcdsaKey{}
	nk.Initialize()
	return nk
}

// NewEcdsaKeyFromPrivate creates and initialize a IHiveKey object from an existing ECDSA private key.
func NewEcdsaKeyFromPrivate(privKey *ecdsa.PrivateKey) IHiveKey {
	k := &EcdsaKey{
		// reuse RSA
		RsaKey{
			privKeyPtr: privKey,
			pubKeyPtr:  &privKey.PublicKey,
		},
	}
	return k
}
